### ISSUE TITLE
with_field on typetracer doesn't touch assigned data (may not be a bug but needs addressing)

### ISSUE THREAD

**Issue by lgray** [scikit-hep/awkward#2431]
### Version of Awkward Array

3.1.4

### Description and code to reproduce

@martindurant @douglasdavis @jpivarski @agoose77 

Related to https://github.com/dask-contrib/dask-awkward/issues/223

This might not be a bug, depending on what is expected (I think below is actually correct, as I understand).
Right now this causes some problems in dask_awkward since `dak.with_field` cannot be optimized away as an operation, so with_field gets called without the data it needs loaded from file, resulting in failure if the data assigned via with_field is not specifically used later on.

I suspect this should be fixed in dask_awkward so those operations can simply be culled.

the following script:
```python3
import json
import awkward as ak
import numpy as np

fromjson = {
    "class": "RecordArray",
    "fields": ["muon", "flatlist"],
    "contents": [
        {
            "class": "ListOffsetArray",
            "offsets": "i64",
            "content": {
                "class": "RecordArray",
                "fields": ["pt", "eta", "phi", "crossref", "thing1"],
                "contents": [
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_pt!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_eta!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_phi!",
                    },
                    {
                        "class": "ListOffsetArray",
                        "offsets": "i64",
                        "content": {
                            "class": "NumpyArray",
                            "primitive": "int64",
                            "inner_shape": [],
                            "parameters": {},
                            "form_key": "muon_crossref_content!",
                        },
                        "parameters": {},
                        "form_key": "muon_crossref_index!",
                    },
                    {
                        "class": "NumpyArray",
                        "primitive": "int64",
                        "inner_shape": [],
                        "parameters": {},
                        "form_key": "muon_thing1!",
                    },
                ],
                "parameters": {},
                "form_key": "muon_record!",
            },
            "parameters": {},
            "form_key": "muon_list_outer!",
        },
        {
            "class": "NumpyArray",
            "primitive": "float64",
            "inner_shape": [],
            "parameters": {},
            "form_key": "flat_list!",
        },
    ],
    "parameters": {},
    "form_key": "outer!",
}

form = ak.forms.from_json(json.dumps(fromjson))

ttlayout, report = ak._nplikes.typetracer.typetracer_with_report(form, forget_length=True)

ttarray = ak.Array(ttlayout)

muons = ttarray.muon

muons["flatlist"] = ttarray.flatlist

print(report.data_touched)

muons.flatlist

print(report.data_touched)
```
prints out:
```
['muon_list_outer!']
['muon_list_outer!']
```

**Comment by agoose77** [comment]
With the view that data touching occurs when an operation meaningfully depends upon the contents of a buffer, I don't expect `with_field` to touch the `flat_list!` key. The proper way for optimisations to be applied, though, is to inject "fake" buffers that allow us to avoid rewriting layouts (which is not just easier to reason about, but is also more correct). That's being tracked [here](https://github.com/dask-contrib/dask-awkward/pull/184). 

I'll ping @jpivarski just to get his perspective; I don't want to make a bold statement that's incorrect here.

If my view holds, then we'll need to move this issue to `dask-awkward` :)

**Comment by agoose77** [comment]
Looking at this again, it's clear that the above statement is wrong: we _should_ touch `flat_list!` because it gets broadcasted to fit the muon list structure! This is also fixed — we now touch these data due to advanced indexing of typetracer arrays during broadcasting. Thanks for the issue Lindsey!

Closed by #2479
